extensions/sql-database-projects/src/controllers/projectController.ts

```diff
@@ -486,6 +494,78 @@ export class ProjectsController {
 		}
 	}
 
+	public async schemaCompareGetTargetScripts(projectFilePath: string): Promise<string[]> {
+		const project = await Project.openProject(projectFilePath);
+		let files: string[] = [];
+
+		const fileEntries: FileProjectEntry[] = project.files;
+
+		fileEntries.forEach(f => {
+
+			if (f.fsUri.fsPath.endsWith('.sql')) {
+				files.push(f.fsUri.fsPath);
+			}
+		});
+
+		return files;
+	}
+
+	public async schemaCompareGetDsp(projectFilePath: string): Promise<string> {
+		const project = await Project.openProject(projectFilePath);
+		return project.getProjectTargetVersion();
+	}
+
+	public async schemaComparePublishProjectChanges(operationId: string, projectFilePath: string, folderStructure: string): Promise<mssql.SchemaComparePublishProjectResult> {
+		let ext = vscode.extensions.getExtension(mssql.extension.name)!;
+		const service = (await ext.activate() as mssql.IExtension).schemaCompare;
+
+		const projectPath = path.dirname(projectFilePath);
+
+		let fs: mssql.ExtractTarget = mssql.ExtractTarget.flat;
+		switch (folderStructure) {
+			case constants.file:
+				fs = mssql.ExtractTarget.file;
+				break;
+			case constants.flat:
+				fs = mssql.ExtractTarget.flat;
+				break;
+			case constants.objectType:
+				fs = mssql.ExtractTarget.objectType;
+				break;
+			case constants.schema:
+				fs = mssql.ExtractTarget.schema;
+				break;
+			case constants.schemaObjectType:
+				fs = mssql.ExtractTarget.schemaObjectType;
+				break;
+		}
+
+		const result: mssql.SchemaComparePublishProjectResult = await service.schemaComparePublishProjectChanges(operationId, projectPath, fs, utils.getAzdataApi()!.TaskExecutionMode.execute);
+
+		const project = await Project.openProject(projectFilePath);
+
+		let toAdd: vscode.Uri[] = [];
+		result.addedFiles.forEach((f: any) => toAdd.push(vscode.Uri.file(f)));
+		await project.addToProject(toAdd);
+
+		let toRemove: vscode.Uri[] = [];
+		result.deletedFiles.forEach((f: any) => toRemove.push(vscode.Uri.file(f)));
+
+		let toRemoveEntries: FileProjectEntry[] = [];
+		toRemove.forEach(f => toRemoveEntries.push(new FileProjectEntry(f, f.path.replace(projectPath + '\\', ''), EntryType.File)));
+
+		toRemoveEntries.forEach(async f => await project.exclude(f));
+
+		await this.buildProject(project);
+
+		return result;
+	}
+
+	public async schemaCompareShowProjectsView() {
```

Comment:
this isn't schema compare specific. Other places in this file also use `workspaceApi.showProjectsView()`, so maybe add a helper function for this that can also be used in those places